src/LagrangianParticleTracking/calculate_particle_tendency_contributions.jl

```diff
@@ -0,0 +1,75 @@
+function calculate_particle_tendency_contributions!(particles, model, active_properties)
+    arch = model.architecture
+
+    workgroup = min(length(particles), 256)
+    worksize = length(particles)
+
+    events = []
+
+    for field in active_properties
+        property = getproperty(particles.properties, field.property)
+        tendency = getproperty(model.timestepper.Gⁿ, field.tracer)
+
+        Gp_kernel! = calculate_particle_tendency_kernel!(device(arch), workgroup, worksize)
+        Gp_event = Gp_kernel!(property, tendency, particles, model.tracers, model.grid)
+
+        push!(events, Gp_event)
+    end
+
+    wait(device(arch), MultiEvent(Tuple(events)))
+
+    return nothing
+end
+
+calculate_particle_tendency_contributions!(model) = calculate_particle_tendency_contributions!(model.particles, model)
+calculate_particle_tendency_contributions!(particles::Nothing, model) = nothing
+calculate_particle_tendency_contributions!(particles, model) = !isnothing(particles.parameters) ? (:active_properties in keys(particles.parameters) ? calculate_particle_tendency_contributions!(particles, model, particles.parameters.active_properties) : nothing) : nothing
+calculate_particle_tendency_contributions!(particles, model, active_fields::Nothing) = nothing
+
+@inline function get_mask(x, y, z, grid::AbstractGrid{FT, TX, TY, TZ}, loc) where {FT, TX, TY, TZ}
+    (fi, i::Int), (fj, j::Int), (fk, k::Int) = modf.(fractional_indices(x, y, z, loc, grid))
+
+    nodesi = arch_array(grid.architecture, zeros(Int, 2, 2, 2))
+    nodesj = arch_array(grid.architecture, zeros(Int, 2, 2, 2))
+    nodesk = arch_array(grid.architecture, zeros(Int, 2, 2, 2))
+    d = arch_array(grid.architecture, zeros(2, 2, 2))
+    
+    if (fi, fj, fk) == (0, 0, 0) #particle on grid point special case
+        @inbounds nodesi[1, 1, 1], nodesj[1, 1, 1], nodesk[1, 1, 1], d = i, j, k, 1
+    else
+        for (a, di) in enumerate([fi, 1-fi]), (b, dj) in enumerate([fj, 1-fj]), (c, dk) in enumerate([fk, 1-fk])
+            @inbounds begin
+                d[a,b,c] = sqrt(di^2+dj^2+dk^2)
+                nodesi[a, b, c] = getnode(TX(), i+a, grid.Nx)
+                nodesj[a, b, c] = getnode(TY(), j+b, grid.Ny)
+                nodesk[a, b, c] = getnode(TZ(), k+c, grid.Nz)
+            end
+        end
+        
+        normfactor=1/sum(1 ./ d)
+        d ./= normfactor
+    end
+
+    return nodesi, nodesj, nodesk, d
+end
```

Comment:
It may reduce the clarity of what this function does too much but I rewrote this function and its about 10x faster:

```suggestion
@inline function get_mask(x, y, z, grid, loc)
    (fi, i::Int), (fj, j::Int), (fk, k::Int) = modf.(fractional_indices(x, y, z, loc, grid))

    if (fi, fj, fk) == (0, 0, 0) #particle on grid point special case
        return ((i, j, k, 1), ), 1.0
    else
        nodes = repeat([(1, 1, 1, NaN)], 8)
        _normfactor = 0.0
        @unroll for n=1:8
            a, b, c = 0^(1+(-1)^n), 0^(1+(-1)^floor(n/2)), 0^(1+(-1)^floor(n/4))
            di, dj, dk = 0^abs(1-a)+fi*(-1)^a, 0^abs(1-b)+fj*(-1)^b, 0^abs(1-c)+fk*(-1)^c #when (e.g. a=0 this makes di, the x distance from the particle to the grid point di, and when a=1 it is 1-fi
            nodes[n] = (i+1+a, 
                            j+1+b, 
                            k+1+c, 
                            sqrt(di^2+dj^2+dk^2))
                            _normfactor += 1 ./sqrt(di^2+dj^2+dk^2)
        end
        return nodes, 1/_normfactor
    end
end
```